chain/lf3/ec.go

```diff
@@ -20,52 +21,56 @@ import (
 	"github.com/filecoin-project/lotus/chain/vm"
 )
 
+var (
+	_ ec.Backend = (*ecWrapper)(nil)
+	_ ec.TipSet  = (*f3TipSet)(nil)
+
+	emptyBeacon = make([]byte, 32)
+)
+
 type ecWrapper struct {
 	ChainStore   *store.ChainStore
+	Syncer       *chain.Syncer
 	StateManager *stmgr.StateManager
 }
 
-var _ ec.TipSet = (*f3TipSet)(nil)
-
-type f3TipSet types.TipSet
-
-func (ts *f3TipSet) cast() *types.TipSet {
-	return (*types.TipSet)(ts)
+type f3TipSet struct {
+	*types.TipSet
 }
 
-func (ts *f3TipSet) String() string {
-	return ts.cast().String()
-}
+func (ts *f3TipSet) String() string       { return ts.TipSet.String() }
+func (ts *f3TipSet) Key() gpbft.TipSetKey { return ts.TipSet.Key().Bytes() }
+func (ts *f3TipSet) Epoch() int64         { return int64(ts.TipSet.Height()) }
 
-func (ts *f3TipSet) Key() gpbft.TipSetKey {
-	return ts.cast().Key().Bytes()
+func (ts *f3TipSet) FirstBlockHeader() *types.BlockHeader {
+	if ts.TipSet == nil || len(ts.TipSet.Blocks()) == 0 {
+		return nil
+	}
+	return ts.TipSet.Blocks()[0]
 }
 
 func (ts *f3TipSet) Beacon() []byte {
-	entries := ts.cast().Blocks()[0].BeaconEntries
-	if len(entries) == 0 {
-		// This should never happen in practice, but set beacon to a non-nil
-		// 32byte slice to force the message builder to generate a
-		// ticket. Otherwise, messages that require ticket, i.e. CONVERGE will fail
-		// validation due to the absence of ticket. This is a convoluted way of doing it.
-		return make([]byte, 32)
-	}
-	return entries[len(entries)-1].Data
-}
-
-func (ts *f3TipSet) Epoch() int64 {
-	return int64(ts.cast().Height())
+	switch header := ts.FirstBlockHeader(); {
+	case header == nil, len(header.BeaconEntries) == 0:
+		// This should never happen in practice, but set beacon to a non-nil 32byte slice
+		// to force the message builder to generate a ticket. Otherwise, messages that
+		// require ticket, i.e. CONVERGE will fail validation due to the absence of
+		// ticket. This is a convoluted way of doing it.
+
+		// TODO: investigate if this is still necessary, or how message builder can be
+		//       adapted to behave correctly regardless of beacon value, e.g. fail fast
+		//       instead of building CONVERGE with empty beacon.
+		return emptyBeacon
```

Comment:
I'd just... re-create this every time. Safer and we shouldn't hit this case anyways.